doc/pcre2_config.3
.TH PCRE2_CONFIG 3 "03 September 2025" "PCRE2 10.49"
.SH NAME
PCRE2 - Perl-compatible regular expressions (revised API)
.SH SYNOPSIS
.rs
.sp
.B #include <pcre2.h>
.PP
.nf
.B int pcre2_config(uint32_t \fIwhat\fP, void *\fIwhere\fP);
.fi
.
.SH DESCRIPTION
.rs
.sp
This function makes it possible for a client program to find out which optional
features are available in the version of the PCRE2 library it is using. The
arguments are as follows:
.sp
  \fIwhat\fP     A code specifying what information is required
  \fIwhere\fP    Points to where to put the information
.sp
If \fIwhere\fP is NULL, the function returns the amount of memory needed for
the requested information. When the information is a string, the value is in
code units; for other types of data it is in bytes.
.P
If \fBwhere\fP is not NULL, for PCRE2_CONFIG_JITTARGET,
PCRE2_CONFIG_UNICODE_VERSION, and PCRE2_CONFIG_VERSION it must point to a
buffer that is large enough to hold the string. For all other codes it must
point to a uint32_t integer variable. The available codes are:
.sp
  PCRE2_CONFIG_BSR                Indicates what \eR matches by default:
                                    PCRE2_BSR_UNICODE
                                    PCRE2_BSR_ANYCRLF
  PCRE2_CONFIG_COMPILED_WIDTHS    Which of 8/16/32 support was compiled
  PCRE2_CONFIG_DEPTHLIMIT         Default backtracking depth limit
  PCRE2_CONFIG_EFFECTIVE_LINKSIZE How many bytes are used for link size
  PCRE2_CONFIG_HEAPLIMIT          Default heap memory limit
.\" JOIN
  PCRE2_CONFIG_JIT                Availability of just-in-time compiler
                                   support (1=yes 0=no)
.\" JOIN
  PCRE2_CONFIG_JITTARGET          Information (a string) about the target
                                   architecture for the JIT compiler
  PCRE2_CONFIG_LINKSIZE           Configured internal link size (2, 3, 4)
  PCRE2_CONFIG_MATCHLIMIT         Default internal resource limit
  PCRE2_CONFIG_NEVER_BACKSLASH_C  Whether or not \eC is disabled
  PCRE2_CONFIG_NEWLINE            Code for the default newline sequence:
                                    PCRE2_NEWLINE_CR
                                    PCRE2_NEWLINE_LF
                                    PCRE2_NEWLINE_CRLF
                                    PCRE2_NEWLINE_ANY
                                    PCRE2_NEWLINE_ANYCRLF
                                    PCRE2_NEWLINE_NUL
  PCRE2_CONFIG_PARENSLIMIT        Default parentheses nesting limit
  PCRE2_CONFIG_RECURSIONLIMIT     Obsolete: use PCRE2_CONFIG_DEPTHLIMIT
  PCRE2_CONFIG_STACKRECURSE       Obsolete: always returns 0
.\" JOIN
  PCRE2_CONFIG_UNICODE            Availability of Unicode support
                                   (1=yes 0=no)
  PCRE2_CONFIG_UNICODE_VERSION    The Unicode version (a string)
  PCRE2_CONFIG_VERSION            The PCRE2 version (a string)
.sp
The function yields a non-negative value on success or the negative value
PCRE2_ERROR_BADOPTION otherwise. This is also the result for the
PCRE2_CONFIG_JITTARGET code if JIT support is not available. When a string is
requested, the function returns the number of code units used, including the
terminating zero.
.P
There is a complete description of the PCRE2 native API in the
.\" HREF
\fBpcre2api\fP
.\"
page and a description of the POSIX API in the
.\" HREF
\fBpcre2posix\fP
.\"
page.


NEWS
News about PCRE2 releases
-------------------------

Version 10.49 28-September-2026
-------------------------------

This is a security-only release, to address GHSA-r9hj-j2rw-4q3m.

Compared to 10.48, this release has only a minimal code change to prevent an
out-of-bounds write with arbitrary data. An attacker-controlled regex pattern
is required, and applications are only affected if using the
pcre2_jit_stack_create() and pcre2_jit_stack_assign() APIs to provide a growable
JIT stack, and then matching against a pattern with unusually high JIT stack
usage, such as a large number of capturing groups.

The implications of an out-of-bounds write could include arbitrary code
execution.

The issue is not a regression and affects releases 10.48 and earlier.


Version 10.48 31-August-2026
----------------------------

This is a regular release, incorporating security fixes along with small
improvements and fixes to library behaviour.

Only changes to behaviour, changes to the API, and other significant changes
are described here. Please see the ChangeLog and Git log for further details.

As well as the tarball and Git tag for this release, there are detailed
instructions for backporting security and correctness fixes, for the last
five years of releases.

* (Git change) Renamed the default development branch from master to main.

* (Maintenance change) Added a five-year support lifecycle policy and
publication of backport patches for security and high-severity fixes in older
releases.

* (Security fix for very specific API usage, GHSA-2p8c-ff85-vh9x) If
pcre2_jit_compile() is called with options for some match modes, and then
pcre2_match() is used to perform a match for a different match mode, an
out-of-bounds read can occur if the match is attempted against invalid UTF input.

* (Security fix for pattern conversion, GHSA-q8g2-wprr-34m9) If pcre2_convert()
is called on untrusted input on platforms with 32-bit size_t, an out-of-bounds
heap write can occur.

* (Security fix, GHSA-3r4p-g7gg-ppmf) Fixed an out-of-bounds write in DFA
matching when using a heap limit; also fixed possible integer overflows which
could cause under-allocation of the workspace.

* (Security fix, GHSA-fmgr-6ggq-9859) Added bounds checks for several integer
overflows while compiling patterns on 32-bit CPUs, which could cause
under-allocation followed by out-of-bounds writes.

* (Security fix, GHSA-9qww-pwc4-77qq) Applied lower buffer bound to prevent
two out-of-bounds reads while scanning backwards through invalid UTF data with
PCRE2_MATCH_INVALID_UTF.

* (Matching correctness) Fixed several matching issues:
  - A JIT-specific matching bug affecting prefix scanning on patterns with
    repeats (#875).
  - A JIT-specific matching bug in variable-length lookbehinds (#912).
  - Miscompiled Unicode character classes combining characters at or below
    U+00FF with characters at U+0100 and U+8000 or above (#841).
  - Incorrect JIT character advancement with PCRE2_MATCH_INVALID_UTF in UTF-8
    and UTF-16 modes, which could skip adjacent characters (#945).

* (Behaviour change) Updated Unicode support to Unicode 17.0.

* (Small behaviour changes) Many small fixes, including pcre2_substitute()
improvements, optimisation of possessive backreference matching, and
pcre2_compile() fixes.

* (Small build changes) Many small adjustments to the CMake and Zig builds.

* (Security fix for very specific API usage, #937) Fixed a leak and later
invalid free when calling the fast-path pcre2_jit_match() function with a match
data object previously used with pcre2_match() and
PCRE2_COPY_MATCHED_SUBJECT.

* (Low-severity security fix, GHSA-q7rw-r7qq-2hx6) Fixed exposure of two
uninitialised bytes from malloc() via pcre2_serialize_encode().


Version 10.47 21-October-2025
-----------------------------

This is a regular semi-annual release, incorporating a few new features and
several maintenance and build improvements.

Only changes to behaviour, changes to the API, and other significant changes
are described here. Please see the ChangeLog and Git log for further details.

* (Powerful new feature) Pattern recursion of the form
"(?1(GROUP_NAME_OR_NUM,...))" acts as a subroutine call which additionally
returns the listed capturing groups to the calling context.

* (Significant bugfix) Fixed a crash in pcre2_callout_enumerate() which is
easily reachable on any pattern that contains a Unicode character class. If your
application uses this function, please read the details for this change and
evaluate its severity for your application.

* (Build change) There are now linker scripts to enable symbol versioning for
the PCRE2 dynamic libraries. Downstream Linux distributions may make use of
this, or disable it with the new Autoconf `--disable-symvers` and CMake
`-DPCRE2_SYMVERS` options. Linux, Solaris, and FreeBSD (GNU ld, LLVM lld, and
Solaris ld) are tested and supported.

* (New API function) Added pcre2_next_match(). This function makes it both
simpler and safer for clients to iterate over all matches in a subject. The
documentation in `pcre2api` also provides improved guidance in the section
"Iterating over all matches".

* (Minor API addition) Added the PCRE2_CONFIG_EFFECTIVE_LINKSIZE option to
pcre2_config().

* (Minor replacement syntax extension) Added support for $+ replacement to
pcre2_substitute().

* (Build change) Modernize the CMake build files, to use the
"$<BUILD_INTERFACE:...>", "$<INSTALL_INTERFACE:...>" and "install(EXPORT...)"
expressions to export the PCRE2 targets.


Version 10.46 27-August-2025
----------------------------

This is a security-only release, to address CVE-2025-58050.

Compared to 10.45, this release has only a minimal code change to prevent a
read-past-the-end memory error, of arbitrary length. An attacker-controlled
regex pattern is required, and it cannot be triggered by providing crafted
subject (match) text. The (*ACCEPT) and (*scs:) pattern features must be used
together.


src/pcre2_config.c
/*************************************************
*      Perl-Compatible Regular Expressions       *
*************************************************/

/* PCRE is a library of functions to support regular expressions whose syntax
and semantics are as close as possible to those of the Perl 5 language.

                       Written by Philip Hazel
     Original API code Copyright (c) 1997-2012 University of Cambridge
          New API code Copyright (c) 2016-2020 University of Cambridge

-----------------------------------------------------------------------------
Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright notice,
      this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.

    * Neither the name of the University of Cambridge nor the names of its
      contributors may be used to endorse or promote products derived from
      this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.
-----------------------------------------------------------------------------
*/


#include "pcre2_internal.h"



/* These macros are the standard way of turning unquoted text into C strings.
They allow macros like PCRE2_MAJOR to be defined without quotes, which is
convenient for user programs that want to test their values. */

#define STRING(a)  #a
#define XSTRING(s) STRING(s)


/*************************************************
* Return info about what features are configured *
*************************************************/

/* If where is NULL, the required size is returned: bytes for numerical values,
or code units including the terminating zero for strings. Otherwise, the
requested information is written to where.

Arguments:
  what             what information is required
  where            where to put the information, or NULL to query its size

Returns:           sizeof(uint32_t) for a numerical value if where is NULL
                   0 if a numerical value is written
                   number of code units including the terminating zero
                     if a string is requested
                   PCRE2_ERROR_BADOPTION if "what" not recognized
                     or JIT target requested when JIT not enabled
*/

PCRE2_EXP_DEFN int PCRE2_CALL_CONVENTION
pcre2_config(uint32_t what, void *where)
{
  if (where == NULL) // Requests a length
  {
    switch (what)
    {
    default:
      return PCRE2_ERROR_BADOPTION;

    case PCRE2_CONFIG_BSR:
    case PCRE2_CONFIG_COMPILED_WIDTHS:
    case PCRE2_CONFIG_DEPTHLIMIT:
    case PCRE2_CONFIG_EFFECTIVE_LINKSIZE:
    case PCRE2_CONFIG_HEAPLIMIT:
    case PCRE2_CONFIG_JIT:
    case PCRE2_CONFIG_LINKSIZE:
    case PCRE2_CONFIG_MATCHLIMIT:
    case PCRE2_CONFIG_NEVER_BACKSLASH_C:
    case PCRE2_CONFIG_NEWLINE:
    case PCRE2_CONFIG_PARENSLIMIT:
    case PCRE2_CONFIG_STACKRECURSE: // Obsolete
    case PCRE2_CONFIG_TABLES_LENGTH:
    case PCRE2_CONFIG_UNICODE:
      return sizeof(uint32_t);

      /* These are handled below */

    case PCRE2_CONFIG_JITTARGET:
    case PCRE2_CONFIG_UNICODE_VERSION:
    case PCRE2_CONFIG_VERSION:
      break;
    }
  }

  switch (what)
  {
  default:
    return PCRE2_ERROR_BADOPTION;

  case PCRE2_CONFIG_BSR:
    *((uint32_t *)where) = BSR_DEFAULT;
    break;

  case PCRE2_CONFIG_COMPILED_WIDTHS:
    *((uint32_t *)where) = 0
#ifdef SUPPORT_PCRE2_8
                           + (1 << 0)
#endif
#ifdef SUPPORT_PCRE2_16
                           + (1 << 1)
#endif
#ifdef SUPPORT_PCRE2_32
                           + (1 << 2)
#endif
        ;
    break;

  case PCRE2_CONFIG_DEPTHLIMIT:
    *((uint32_t *)where) = MATCH_LIMIT_DEPTH;
    break;

  case PCRE2_CONFIG_EFFECTIVE_LINKSIZE:
    *((uint32_t *)where) = LINK_SIZE * sizeof(PCRE2_UCHAR);
    break;

  case PCRE2_CONFIG_HEAPLIMIT:
    *((uint32_t *)where) = HEAP_LIMIT;
    break;

  case PCRE2_CONFIG_JIT:
#ifdef SUPPORT_JIT
    *((uint32_t *)where) = 1;
#else
    *((uint32_t *)where) = 0;
#endif
    break;

  case PCRE2_CONFIG_JITTARGET:
#ifdef SUPPORT_JIT
    {
      const char *v = PRIV(jit_get_target)();
      return (int)(1 + ((where == NULL) ? strlen(v) : PRIV(strcpy_c8)((PCRE2_UCHAR *)where, v)));
    }
#else
    return PCRE2_ERROR_BADOPTION;
#endif

  case PCRE2_CONFIG_LINKSIZE:
    *((uint32_t *)where) = (uint32_t)CONFIGURED_LINK_SIZE;
    break;

  case PCRE2_CONFIG_MATCHLIMIT:
    *((uint32_t *)where) = MATCH_LIMIT;
    break;

  case PCRE2_CONFIG_NEWLINE:
    *((uint32_t *)where) = NEWLINE_DEFAULT;
    break;

  case PCRE2_CONFIG_NEVER_BACKSLASH_C:
#ifdef NEVER_BACKSLASH_C
    *((uint32_t *)where) = 1;
#else
    *((uint32_t *)where) = 0;
#endif
    break;

  case PCRE2_CONFIG_PARENSLIMIT:
    *((uint32_t *)where) = PARENS_NEST_LIMIT;
    break;

    /* This is now obsolete. The stack is no longer used via recursion for
    handling backtracking in pcre2_match(). */

  case PCRE2_CONFIG_STACKRECURSE:
    *((uint32_t *)where) = 0;
    break;

  case PCRE2_CONFIG_TABLES_LENGTH:
    *((uint32_t *)where) = TABLES_LENGTH;
    break;

  case PCRE2_CONFIG_UNICODE_VERSION:
    {
#if defined SUPPORT_UNICODE
      const char *v = PRIV(unicode_version);
#else
      const char *v = "Unicode not supported";
#endif
      return (int)(1 + ((where == NULL) ? strlen(v) : PRIV(strcpy_c8)((PCRE2_UCHAR *)where, v)));
    }

  case PCRE2_CONFIG_UNICODE:
#if defined SUPPORT_UNICODE
    *((uint32_t *)where) = 1;
#else
    *((uint32_t *)where) = 0;
#endif
    break;

  case PCRE2_CONFIG_VERSION:
    {
      /* Note that XSTRING(PCRE2_PRERELEASE) attempts to apply the STRING() macro to an empty sequence
      of tokens in release builds (when PCRE2_PRERELEASE is empty). This is valid as of C99, but
      historically some compilers did not support this. */

      const char *v = XSTRING(PCRE2_MAJOR) "." XSTRING(PCRE2_MINOR) XSTRING(PCRE2_PRERELEASE) " " XSTRING(PCRE2_DATE);
      return (int)(1 + ((where == NULL) ? strlen(v) : PRIV(strcpy_c8)((PCRE2_UCHAR *)where, v)));
    }
  }

  return 0;
}

/* End of pcre2_config.c */
